### PR TITLE
Fixed #27453 -- Avoided unnecessary recompilation of non-translated URL regexes.

### PR DESCRIPTION
**Pull request by carljm**

### CODE REVIEW

File: django/urls/resolvers.py [hunk 1]
```diff
@@ -90,6 +90,11 @@ def __init__(self, regex):
         # expression.
         self._regex = regex
         self._regex_dict = {}
+        # As a performance optimization, if the given regex string is a regular
+        # string (not a lazily-translated string proxy), we can just compile it
+        # once and avoid the property and the per-langage compilation entirely.
```
Comment (apollo13) [inline django/urls/resolvers.py]: `per-language`

File: django/urls/resolvers.py [hunk 2]
```diff
@@ -90,6 +90,11 @@ def __init__(self, regex):
...
+        if isinstance(regex, six.string_types):
+            self.__dict__['regex'] = self._compile(regex)
```
Comment (knbk) [inline django/urls/resolvers.py]: Setting `self.__dict__['regex']` doesn't work with `@property` (tested on 2.7). I don't think it's even possible to override the return value of a property like this. You'll have to dynamically return the right value in the `regex` getter.

Comment (charettes) [inline django/urls/resolvers.py]: It doesn't work with `property` as it defines a `__set__` method. Using a readonly descriptor (only defining `__get__` à la `cached_property`) should work here and be worth it as it would prevent a function call in what seems to a be a performance critical path.

Comment (apollo13) [inline django/urls/resolvers.py]: Yeah, we will need a testcase either way ;)

File: tests/urlpatterns_reverse/translations/locale/fr/formats.py [hunk 3]
```diff
@@ -0,0 +1,2 @@
+# A user-defined format
+CUSTOM_DAY_FORMAT = 'd/m/Y CUSTOM'
```
Comment (knbk) [inline tests/urlpatterns_reverse/translations/locale/fr/formats.py]: Is this file necessary? It doesn't seem to be used anywhere.

Comment (carljm) [inline tests/urlpatterns_reverse/translations/locale/fr/formats.py]: Argh, thanks for catching that. Copied-directory foul, will remove.

**Comment by carljm** [comment]
PR updated to remove stray `formats.py` file in tests.

**Comment by timgraham** [comment]
Looks okay to me. I split it into two commits since some of the tests already pass before the fix. I also fixed a few other small things such as allowing the tests to run on non-ASCII paths for Python 2 (`upath` addition).

@claudep, if you have a chance, please let us know if you have any input on this.

**Comment by claudep** [comment]
That's fine, thanks.
